### PR TITLE
Core/cleanup tests

### PR DESCRIPTION
**Pull request by philbucher**
This PR cleans deletes files created by the restart and the gidio test, based on @jcotela suggestion.
I also used the opportunity to:
- Add a testing_utilities file for common functionalities (so far only file deleting)
- Add a runscript for the cpp tests and add it to the test_KratosCore.py runner (I hope this is ok, I took the idea from the hdf5 app). I coulnd't figure out a way to run the tests from the Core only, can you help me there @roigcarlo please?
- Made the `GetFilePath` function that is used in many tests independent of the OS

### CODE REVIEW

File: kratos/tests/run_cpp_unit_tests_KratosCore.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from KratosMultiphysics import *
+
+def runTests():
+    Tester.SetVerbosity(Tester.Verbosity.TESTS_OUTPUTS)
+    Tester.RunTestCases("KratosCore*") # Is there a way to detect the cpp tests in the core?
```
Comment (philbucher) [inline kratos/tests/run_cpp_unit_tests_KratosCore.py]: @roigcarlo is there a way to add the cpp tests of the Core only?

**Comment by jcotela** [comment]
@philbucher I think that @roigcarlo already did some work to ensure that the core cpp tests are run when you launch the python test runner. Regarding selecting the tests to run, I think the plan was to use different suites for the different applications (in fluids we have a FluidDynamicsFastSuite or something like that). In any case, I'd be tempted to always run all cpp tests in the fast suite with the runner (not just the core). In this way, we all have an interest in keeping the tests on the main run queue very short.

**Comment by roigcarlo** [comment]
@philbucher as @jcotela says the cpp tests already run with the python ones. It works as follows:

- __python_scripts/run_tests.py__: This file is the 'main' file for testing, by default it tests KratosCore python tests and tries to run tests for all the applications it finds in the kratos install directory. At the end, it runs all the cpp tests for the core (but not for the applications).

- __python_scripts/run_cpp_tests.py__: This one only runs the cpp tests for the core.

- __application/tests/test_XXXXXX.py__: These ones are the main python suits for the application/core. I think these should not run the cpp tests, as are different things. For example you are requiring an application to have python tests so it can run the cpp ones (conceptually speaking).

For what I can see, what you are trying to do is a python script that run only the cpp tests of the core, which basically is the default behavior of `Tester.RunAllTestCases()`. The trick here is that only the tests of the modules loaded in python when the function is called will run, so `python_scripts/run_cpp_tests.py` should be enough. Example:

```Python
from KratosMultiphysics import *

Tester.RunAllTestCases() # This will run ONLY KratosCore cpp tests 
```

```Python
from KratosMultiphysics import *
from KratosMultiphysics.FluidDynamicsApplication import *

Tester.RunAllTestCases() # This will run Both KratosCore and FluidDynamicsApplication cpp tests
```

**Comment by philbucher** [comment]
Thx for your answers and explanations!
I reverted the unwanted changes

File: kratos/python_scripts/testing_utilities.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+# Other imports
+import os
+
+def DeleteFileIfExisting(file_name):
+    if rank == 0:
+        try:
+            os.remove(file_name)
+        except FileNotFoundError:
+            pass
```
Comment (msandre) [inline kratos/python_scripts/testing_utilities.py]: os.remove can also throw IsADirectoryError.

Comment (msandre) [reply kratos/python_scripts/testing_utilities.py]: You can also use os.path.isfile(file_name).

File: kratos/tests/test_KratosCore.py [hunk 3]
```diff
@@ -28,7 +28,7 @@
 import test_gid_io_gauss_points
 
 
-def AssambleTestSuites():
+def AssembleTestSuites():
```
Comment (msandre) [inline kratos/tests/test_KratosCore.py:31]: :+1:

File: kratos/python_scripts/testing_utilities.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+from __future__ import print_function, absolute_import, division
+
+# Import Kratos
+import KratosMultiphysics
+try:
+    import KratosMultiphysics.mpi as KratosMPI
+    rank = KratosMPI.mpi.rank
```
Comment (msandre) [inline kratos/python_scripts/testing_utilities.py]: Is this safe? I tried this out in interactive python and I get an error code 2 which apparently means MPI_ERR_COUNT.

File: kratos/tests/test_gid_io.py [hunk 5]
```diff
@@ -3,10 +3,12 @@
 import KratosMultiphysics.KratosUnittest as KratosUnittest
 import KratosMultiphysics
 from gid_output_process import GiDOutputProcess
+
+import testing_utilities as test_utils
```
Comment (msandre) [inline kratos/tests/test_gid_io.py]: Is it doable to have testing_utilities as a sub module of KratosUnittest so we only import testing once?

**Review by msandre — COMMENTED** [review]
This is a good improvement. Such functions cut across a lot of test suites (including in adjoint application) and currently appear in various forms. My main question is, can it be a submodule of the kratos unit test? It seems natural to me that it would.

File: kratos/python_scripts/testing_utilities.py (hunk 4 above)
Comment (philbucher) [reply kratos/python_scripts/testing_utilities.py]: I thought it works
At least I use sth similar in [the MappingTests](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/MappingApplication/tests/KratosExecuteMapperTests.py#L35-L46) for a long time now

I still removed it bcs I put the function to the unittest and I didn't want to introduce the dependencie on mpi. If it is there at some point we can add it

File: kratos/tests/test_gid_io.py (hunk 5 above)
Comment (philbucher) [reply kratos/tests/test_gid_io.py]: done

File: kratos/python_scripts/testing_utilities.py (hunk 2 above)
Comment (philbucher) [reply kratos/python_scripts/testing_utilities.py]: thx for the hint

**Comment by philbucher** [comment]
ping

File: kratos/python_interface/kratos_unittest.py [hunk 6]
```diff
@@ -142,6 +142,12 @@ def runTests(tests):
         result = not TextTestRunner(verbosity=verbosity, buffer=True).run(tests[level]).wasSuccessful()
         sys.exit(result)
 
+def DeleteFileIfExisting(file_name):
+    if os.path.isfile(file_name):
+        os.remove(file_name)
+    else:
+        pass
```
Comment (msandre) [inline kratos/python_interface/kratos_unittest.py]: You don't need the else in this case.

**Review by msandre — DISMISSED** [review]

File: kratos/python_interface/kratos_unittest.py (hunk 6 above)
Comment (philbucher) [reply kratos/python_interface/kratos_unittest.py]: thx for the hint!

**Review by msandre — DISMISSED** [review]

**Comment by pooyan-dadvand** [comment]
@roigcarlo what is your opinion the idea of @msandre about sub modules?

**Comment by roigcarlo** [comment]
I also think it is a great idea. In fact I already use sub modules similar to what @msandre is suggesting for example in the app generator.

I was thinking that maybe if the utilities that we want to add there are generic enough it is worth for that submodule to hang from `KratosMultiphysics` instead of `KratosUnittest` so it can be use by by unittests and pretty much every one else in their own scripts, call it `Kratos_utilities`

**Comment by philbucher** [comment]
@roigcarlo can you suggest or make a template for such a submodule? I could then fill it with the function
I don't know how exactly you want to do it
Thx

**Comment by philbucher** [comment]
ping @roigcarlo 

I saw that there is a [kratos_utilities](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_interface/kratos_utilities.py) existing already but I didn't manage to add functions to it

**Comment by roigcarlo** [comment]
Sorry didn't saw your last message, let me take a look

**Comment by roigcarlo** [comment]
@philbucher You need to "install" kratos after you make changes to that file, it gets moved to KratosMultiphysics folder

**Comment by philbucher** [comment]
@roigcarlo I managed now to add the function to the kratos_utilities, I think this is a good place for it.

Also I added the mpi stuff (file removal only on rank 0) after talking to @jcotela

**Comment by philbucher** [comment]
@roigcarlo can we merge this soon I would like to use features from it in another PR?

**Review by roigcarlo — APPROVED** [review]
Sure, looks good to me

**Comment by philbucher** [comment]
cool thx

**Pull request merged**